### PR TITLE
pylint -> flake8

### PR DESCRIPTION
**Pull request by joelgrus**
this is enormous but shallow, it touches almost every file in the repo 😬

the majority of changes are

* fixing vertical whitespace to be in line with PEP-8 (exactly 2 blank lines before / after a class or bare function; exactly 1 line before / after a class method)
* removing `# pylint: disable=`s (there are a ton of them)
* fixing horizontal whitespace for a variety of reasons

for the most part I approved of all the corrections it wanted (I am not 100% sure about the way it wants dicts formatted over multiple lines, but I don't dislike it)

--

while I was working on this, a new mypy (0.730) came out, so I had to make a few fixes for that as well.

### CODE REVIEW

File: allennlp/common/testing/model_test_case.py [hunk 1]
```diff
@@ -18,7 +18,7 @@ class ModelTestCase(AllenNlpTestCase):
     with added methods for testing :class:`~allennlp.models.model.Model` subclasses.
     """
     def set_up_model(self, param_file, dataset_file):
-        # pylint: disable=attribute-defined-outside-init
```
[the remark this answers is not in the document]
Comment (joelgrus) [reply allennlp/common/testing/model_test_case.py]: I don't think so, it's not a type error per se, it's more of a readability issue

**Comment by joelgrus** [comment]
yeah, I mean, the tradeoff here (as I understand it) is that `flake8` is looking for (1) things that are violations of PEP8, and (2) things that are possibly errors; whereas `pylint` is looking for an immense grab-bag of rules many of which were arbitrarily chosen by the pylint developers.

flake8 does have some plugins that can handle other cases, not sure about these.

**Comment by matt-gardner** [comment]
Yeah, I'd probably say that we should rely on mypy and tests for correctness checks, and have the linter just look at style / formatting (and I've warmed up to the idea of an autoformatter, if we want to go that route).  It's incredibly rare that pylint ever actually found a correctness issue that wouldn't have been caught by something else.

**Comment by joelgrus** [comment]
well, I can construct an alternative `black` PR and we can see what that looks like

**Comment by matt-gardner** [comment]
Are black and flake8 in conflict with each other?

**Comment by joelgrus** [comment]
they don't care about the exact same things; for example, `black` produces

```
            input_gate = torch.sigmoid(
                projected_input[:, 0 * self.hidden_size : 1 * self.hidden_size]
                + projected_state[:, 0 * self.hidden_size : 1 * self.hidden_size]
            )
```

but flake8 doesn't like the spaces around the `:` operator

**Comment by joelgrus** [comment]
FYI, here's a diff after (1) removing all the pylint directives, and (2) running black

https://github.com/allenai/allennlp/compare/master...joelgrus:black?expand=1

**Comment by joelgrus** [comment]
to sum up, (as I understand it), black cares mostly that your code is _formatted the way black likes to format code_; flake8 cares mostly that your code is _idiomatic as per PEP8_. those are overlapping but distinct desiderata.

**Comment by matt-gardner** [comment]
Looking over the `black` diff, there was only one change that I really didn't like, and that's when it puts arguments on a new line, but then has multiple arguments in the line (e.g., https://github.com/allenai/allennlp/compare/master...joelgrus:black?expand=1#diff-434057be418752d64d2760d6c1c75287R334).  Everything else I would be ok with, and it would save a bunch of headache for contributors.

**Comment by joelgrus** [comment]
when I run black on the result of *this* PR, I get two classes of errors:

```
ignore =
    # black doesn't respect "no spaces before :"
    E203
    # or "line break before binary operator"
    W503
```

both of these are explicitly called out in the black README as "things that flake8 doesn't like about black's output, and that it's safe to ignore because they're not actually violations of PEP-8"

based on this, my recommendation is that we merge this PR, and follow up with a PR to introduce black as well. let's discuss at the team meeting today.

**Comment by schmmd** [comment]
@joelgrus this is awesome and a huge step forward!

**Comment by schmmd** [comment]
```
time ./scripts/verify.py --checks pylint                                                                                                                                                                    pylint checks passed
      177.25 real       561.14 user        14.59 sys
```

```
time ./scripts/verify.py --checks flake8
       38.25 real        57.15 user         1.36 sys
```

Love it!

**Pull request merged**